### ISSUE TITLE
Server fails to shutdown due to handle_event_count

### ISSUE THREAD

**Issue by withnale** [sensu/sensu#916]
There are some cirumstances where sensu server is unable to shutdown, which seems to be due to @handling_event_count

At present the exit ruby code in `complete_event_handling` looks like this:

```ruby
retry_until_true do
  if @handling_event_count == 0
    callback.call
    true
  end
end
```

The value is incrememted in `process_event` for every event...

```ruby
handlers.each do |handler|
  @handling_event_count += 1
  filter_event(handler, event) do |event|
    mutate_event(handler, event) do |event_data|
      handle_event(handler, event_data)
    end
  end
end
```

However, following through the logic in `filter_event` there seems to be a number of code paths that result in this value never getting decremented, and consequently when this occurs the standard server shutdown/restart logic is failing.

**Comment by withnale** [comment]
After comparing with 0.16.0 it seems that the tracking of 'handlers in progress' happens after the filters have been applied in the earlier version.

**Comment by withnale** [comment]
This is fixed by the above PR and fixed in 0.17.2
